src/ompl/tools/experience/ExperienceSetup.h

```diff
@@ -0,0 +1,235 @@
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+/* Author: Dave Coleman
+ */
+
+#ifndef OMPL_TOOLS_EXPERIENCE__EXPERIENCE_SETUP_
+#define OMPL_TOOLS_EXPERIENCE__EXPERIENCE_SETUP_
+
+#include "ompl/geometric/SimpleSetup.h"
+
+namespace ompl
+{
+
+    namespace tools
+    {
+
+        /// @cond IGNORE
+        OMPL_CLASS_FORWARD(ExperienceSetup);
+        /// @endcond
+
+        /** \class ompl::geometric::ExperienceSetupPtr
+            \brief A boost shared pointer wrapper for ompl::geometric::ExperienceSetup */
+
+        /** \brief Create the set of classes typically needed to solve a
+            geometric problem */
+        class ExperienceSetup : public ompl::geometric::SimpleSetup
+        {
+        public:
+            /**
+             * \brief Simple logging functionality encapsled in a struct
+             */
+            struct ExperienceStats
+            {
+                ExperienceStats()
+                    : numSolutionsFromRecall_(0)
+                    , numSolutionsFromRecallSaved_(0)
+                    , numSolutionsFromScratch_(0)
+                    , numSolutionsFailed_(0)
+                    , numSolutionsTimedout_(0)
+                    , numSolutionsApproximate_(0)
+                    , numSolutionsTooShort_(0)
+                    , numProblems_(0)
+                    , totalPlanningTime_(0)
+                    , totalInsertionTime_(0)
+                {
+                }
+
+                double getAveragePlanningTime() const
+                {
+                    return totalPlanningTime_ / numProblems_;
+                }
+
+                double getAverageInsertionTime() const
+                {
+                    // Cleanup output
+                    double time = totalInsertionTime_ / numProblems_;
+                    if (time < 0.00000001)
+                        return 0;
+                    else
+                        return totalInsertionTime_ / numProblems_;
+                }
+
+                double numSolutionsFromRecall_;
+                double numSolutionsFromRecallSaved_;
+                double numSolutionsFromScratch_;
+                double numSolutionsFailed_;                
+                double numSolutionsTimedout_;
+                double numSolutionsApproximate_;
+                double numSolutionsTooShort_; // less than 3 states
+                double numProblems_; // input requests
+                double totalPlanningTime_; // of all input requests, used for averaging
+                double totalInsertionTime_; // of all input requests, used for averaging
+            };
+
+            /**
+             * \brief Single entry for the csv data logging file
+             */
+            struct ExperienceLog
```

Comment:
Ability to output to CSV all the statistics of an experience database planner - how often it uses a recalled motion vs. planning from scratch, other other statistics